### PR TITLE
feat: HIP-904 Reject Tokens System Contract implementation

### LINKED ISSUE #15591: HIP-904: Token Reject System Contract
Implement the Token Reject System Contract described in https://github.com/hashgraph/hedera-services/pull/15435

- [x] HSCS System Contract Classes
- [x] Feature flag
- [x] E2E tests

Included E2E test cases:
Positive:
- [x] Verify that the `rejectTokens` function rejects fungible token for an account.
- [x] Verify that the `rejectTokens` function rejects nft token for an account.
- [x] Verify that the `rejectTokens` function rejects multiple tokens for an account.
- [x] Verify that the `rejectTokenFT` function rejects tokens for a given account.
- [x] Verify that the `rejectTokenNFTs` function rejects tokens for a given account and serial number.
- [x] Verify that the `rejectTokenNFTs` function rejects 10 nft serials for a given account.

Negative:
- [x] Verify that the `rejectTokens` function fails when the sender does not have any associated tokens.
- [x] Verify that the `rejectTokens` function fails when the provided fungible token is invalid.
- [x] Verify that the `rejectTokens` function fails when the provided nft is invalid.
- [x] Verify that the `rejectTokens` function fails when 11 or more tokens are provided.
- [x] Verify that the `rejectTokenFT` function fails when the sender does not have any tokens.
- [x] Verify that the `rejectTokensNFT` function fails when the sender does not have any tokens.
- [x] Verify that the `rejectTokensNFT` function fails when 11 or more serials are provided.

### PR DESCRIPTION
**Pull request by stoyanov-st**
**Description**:
This PR aims to implement the Reject Tokens system contract described in https://github.com/hashgraph/hedera-services/pull/15435/files?short_path=03ab1e7#diff-03ab1e73f0ded9a04739c3b1a68c13f779dce416cd2f7b6637b03d8e2e41e105

**Related issue(s)**:

Fixes #15591 

**Notes for reviewer**:
This feature is only enabled for dev and previewnet envs

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java [hunk 1]
```diff
@@ -61,7 +61,8 @@ public enum DispatchType {
     TOKEN_INFO(HederaFunctionality.TOKEN_GET_INFO, DEFAULT),
     UPDATE_TOKEN_CUSTOM_FEES(HederaFunctionality.TOKEN_FEE_SCHEDULE_UPDATE, DEFAULT),
     TOKEN_AIRDROP(HederaFunctionality.TOKEN_AIRDROP, DEFAULT),
-    TOKEN_CLAIM_AIRDROP(HederaFunctionality.TOKEN_CLAIM_AIRDROP, DEFAULT);
+    TOKEN_CLAIM_AIRDROP(HederaFunctionality.TOKEN_CLAIM_AIRDROP, DEFAULT),
+    TOKEN_REJECT(HederaFunctionality.TOKEN_REJECT, TOKEN_FUNGIBLE_COMMON);
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java:65]: (per discussion elsewhere:). Probably needs to have two entries fungible + NF (like MINT/BURN/WIPE above)

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslator.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        this.decoder = decoder;
+    }
+
+    @Override
+    public boolean matches(@NonNull final HtsCallAttempt attempt) {
+        final var rejectEnabled =
+                attempt.configuration().getConfigData(ContractsConfig.class).systemContractRejectTokensEnabled();
+        return attempt.isTokenRedirect()
+                ? attempt.isSelectorIfConfigEnabled(HRC_TOKEN_REJECT_FT, rejectEnabled)
+                        || attempt.isSelectorIfConfigEnabled(HRC_TOKEN_REJECT_NFT, rejectEnabled)
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslator.java:53]: You now have a new method to do both these selectors in one call!

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoderTest.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        // given:
+        given(attempt.configuration()).willReturn(configuration);
+        given(configuration.getConfigData(LedgerConfig.class)).willReturn(ledgerConfig);
+        given(ledgerConfig.tokenRejectsMaxLen()).willReturn(10);
+        given(addressIdConverter.convert(asHeadlongAddress(SENDER_ID.accountNum())))
+                .willReturn(SENDER_ID);
+
+        final var encoded = Bytes.wrapByteBuffer(RejectTokensTranslator.TOKEN_REJECT.encodeCall(
+                Tuple.of(asHeadlongAddress(SENDER_ID.accountNum()), new Address[] {}, new Tuple[] {
+                    Tuple.of(FUNGIBLE_TOKEN_HEADLONG_ADDRESS, 1L)
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoderTest.java:124]: Looks funny in this test to use `FUNGIBLE_TOKEN_{ID,HEADLONG_ADDRESS}` - isn't there a different constant you can use?  (Obviously doesn't matter when you _execute_ this unit test, but it's odd to see.)

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoder.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import com.hedera.node.config.data.LedgerConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class RejectTokensDecoder {
+
+    // Tuple indexes
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoder.java:39]: Do as you did for `Claim` with comments for the different indices/methods?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/TokenRejectSystemContractTest.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    @HapiTest
+    @DisplayName("Reject fungible token")
+    public Stream<DynamicTest> tokenRejectSystemContractTest(@FungibleToken SpecFungibleToken token) {
+        return hapiTest(withOpContext((spec, opLog) -> {
+            allRunFor(spec, sender.associateTokens(token), token.treasury().transferUnitsTo(sender, 100, token));
+            final var tokenAddress = token.addressOn(spec.targetNetworkOrThrow());
+            allRunFor(
+                    spec,
+                    tokenReject.call("rejectTokens", sender, new Address[] {tokenAddress}, new Address[0]),
+                    sender.getBalance().andAssert(balance -> balance.hasTokenBalance(token.name(), 0L)));
```
Comment (lukelee-sl) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/TokenRejectSystemContractTest.java:71]: would it make sense to assert if the token treasury has increase by the number of rejected tokens?

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoderTest.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+                    Tuple.of(FUNGIBLE_TOKEN_HEADLONG_ADDRESS, 1L),
+                    Tuple.of(FUNGIBLE_TOKEN_HEADLONG_ADDRESS, 1L),
+                    Tuple.of(FUNGIBLE_TOKEN_HEADLONG_ADDRESS, 1L),
+                })));
+
+        // when
+        given(attempt.inputBytes()).willReturn(encoded.toArrayUnsafe());
+
+        // then
+        assertThrows(HandleException.class, () -> subject.decodeTokenRejects(attempt));
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoderTest.java:178]: Consider the more complete asserts as for the token cancel PR

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslatorTest.java [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+    private RejectTokensTranslator subject;
+
+    @BeforeEach
+    void setUp() {
+        subject = new RejectTokensTranslator(decoder);
+    }
+
+    @Test
+    void matchesHTSWithConfigEnabled() {
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslatorTest.java:97]: Consider a negative test with invalid selector.

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslatorTest.java [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+                addressIdConverter,
+                verificationStrategies,
+                gasCalculator,
+                configuration);
+
+        // when:
+        var call = subject.callFrom(attempt);
+
+        // then:
+        assertEquals(DispatchForResponseCodeHtsCall.class, call.getClass());
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslatorTest.java:278]: Consider the `verify` condition as in the token cancel PR

**Review by david-bakin-sl — APPROVED** [review]
No problems, but improvements could be made to unit tests similarly to the token cancel PR.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/TokenRejectSystemContractTest.java (hunk 5 above)
Comment (stoyanov-st) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/TokenRejectSystemContractTest.java]: Yes, added treasury balance checks

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java (hunk 1 above)
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/DispatchType.java]: Added gas calculation based on selector.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslator.java (hunk 2 above)
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslator.java]: Yep, done

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoder.java (hunk 4 above)
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoder.java:39]: Done

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoderTest.java (hunk 3 above)
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoderTest.java]: Yep, changed to NFT address

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoderTest.java (hunk 6 above)
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensDecoderTest.java]: Done

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslatorTest.java (hunk 7 above)
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslatorTest.java:123]: Test Added

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslatorTest.java (hunk 8 above)
Comment (stoyanov-st) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/rejecttokens/RejectTokensTranslatorTest.java:334]: Added

**Review by lukelee-sl — APPROVED** [review]
LGTM

**Pull request merged**